dart/dynamics/BodyNode.cpp

```diff
@@ -896,6 +819,60 @@ const Joint* BodyNode::getChildJoint(size_t _index) const
 }
 
 //==============================================================================
+ShapeNode* BodyNode::createShapeNode(const ShapePtr& shape,
+                                     const std::string& name)
+{
+  ShapeNode::Properties properties;
+  properties.mShape = shape;
+
+  return createNode<ShapeNode>(properties);
+}
+
+//==============================================================================
+ShapeNode* BodyNode::createShapeNode(const ShapePtr& shape, const char* name)
+{
+  return createShapeNode(shape, std::string(name));
+}
+
+//==============================================================================
+size_t BodyNode::getNumShapeNodes() const
+{
+  return getNumNodes<ShapeNode>();
+}
+
+//==============================================================================
+ShapeNode* BodyNode::getShapeNode(size_t index)
+{
+  return getNode<ShapeNode>(index);
+}
+
+//==============================================================================
+const ShapeNode* BodyNode::getShapeNode(size_t index) const
+{
+  return getNode<ShapeNode>(index);
+}
+
+//==============================================================================
+std::vector<ShapeNode*> BodyNode::getShapeNodes()
+{
+  std::vector<ShapeNode*> shapeNodes(getNumShapeNodes());
+
+  auto numShapeNode = getNumShapeNodes();
+  for (auto i = 0u; i < numShapeNode; ++i)
+    shapeNodes[i] = getShapeNode(i);
+
+  return shapeNodes;
+}
```

Comment:
> This implementation is dangerous. One std::vector<ShapeNode*> be used for all BodyNode instances.

True. 



> The only advantage of returning a const reference is to avoid creating a new vector in the first place.

I thought returning the `vector` by value would create the `vector` not only the first time but also whenever it gets called. Is it incorrect?

It seems we need an efficient way to get the list of nodes from a node manager.  For now, we should return it by value or have the `vector` for particular nodes in a derived class to return it as a const reference. Or, we just use getNumNode/getNode(index) fashion. I created an issue about this #613, but that wouldn't for this pull request at least.

I think making this function to return `vector` by value would be fine for now.